lib/Conversion/ArcToLLVM/LowerArcToLLVM.cpp

```diff
@@ -173,6 +173,73 @@ struct StorageGetOpLowering : public OpConversionPattern<arc::StorageGetOp> {
   }
 };
 
+struct AssertOpLowering : public OpConversionPattern<arc::AssertOp> {
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(arc::AssertOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+
+    rewriter.replaceOpWithNewOp<scf::IfOp>(
+        op, adaptor.getProperty(),
+        [&](auto &builder, auto loc) {
+          // if property is true then assert will not execute
+          builder.template create<scf::YieldOp>(loc);
+        },
+        [&](auto &builder, auto loc) {
+          // If the condition doesn't hold then print the message and abort
+          auto message = op.getMsg().has_value() ? op.getMsg().value() : "";
```

Comment:
I think there exists a member function of `optional` called `value_or` which would simplify this.